EpikCord/client/websocket_client.py

```diff
@@ -48,198 +58,231 @@ def __init__(
-            )
+            await self.send_json({
+                "op": GatewayOpcode.HEARTBEAT,
+                "d": self.sequence,
+            })
+            return
 
-        if self.interval:
-            await self.send_json(
-                {"op": GatewayOpcode.HEARTBEAT, "d": self.sequence or "null"}
-            )
-            self.heartbeat_time = perf_counter_ns()
-            await asyncio.sleep(self.interval / 1000)
-            logger.debug("Sent a heartbeat!")
+        await asyncio.sleep(self.heartbeat_interval) # type: ignore
+        await self.send_json({
+            "op": GatewayOpcode.HEARTBEAT,
+            "d": self.sequence,
+        })
 
-    async def request_guild_members(
+    async def connect(self):
+        if not self.gateway_url:
+            self.gateway_url = await self.http.get_gateway()["url"]
+
+        self.websocket = await self.http.ws_connect(f"{self.gateway_url}?v=10&encoding=json&compress=zlib-stream")
+        self._closed = False
+        
+        async for event in self.websocket:
+            event_data = event.json()
+            logger.debug(f"Received {event_data} from the Websocket Connection to Discord.")
+
+            if event_data["op"] == GatewayOpcode.DISPATCH:
+                self.sequence = event_data["s"]
+                await self.handle_event(event_data["t"], event_data["d"])
+
+            elif event_data["op"] == GatewayOpcode.HEARTBEAT:
+                await self.heartbeat(True)
+
+            elif event_data["op"] == GatewayOpcode.RECONNECT:
+                await self.reconnect()
+                await self.resume()
+                return
+
+            elif event_data["op"] == GatewayOpcode.INVALID_SESSION:
+                if event_data["d"] == True:
+                    await self.reconnect()
+                    await self.resume()
+                    return
+                await self.reconnect()
+                return
+
+            elif event_data["op"] == GatewayOpcode.HELLO:
+                self.heartbeat_interval = event_data["d"]["heartbeat_interval"] / 1000
+                await self.identify()
+            
+            elif event_data["op"] == GatewayOpcode.HEARTBEAT_ACK:
+                self.heartbeats.append(event_data)
+
+    async def reconnect(self):
+        await self.close()
+        await self.connect()
+
+    async def resume(self):
+        await self.send_json({
+            "op": GatewayOpcode.RESUME,
+            "d": {
+                "token": self.token,
+                "session_id": self.session_id,
+                "seq": self.sequence,
+            },
+        })
+
+    async def handle_event(self, event_name: str, data: Dict):
+        if callback := getattr(self, f"_{event_name}", None):
+            await callback(self, data)
+        
+        for wait_for_callback in self.wait_for_events[event_name]:
+            if await wait_for_callback[1](data):
```

Comment:
Not a fan of the `if await`, i think it would be better leaving `await` in their own line to make it clear that those function might cause a pause in the execution. 

```py
x = await ...

if x:
   ...
```

Reply:
Will do.